scene/3d/chain_ik_3d.cpp

```diff
@@ -0,0 +1,484 @@
+				set_end_bone(which, p_value);
+			} else if (opt == "direction") {
+				set_end_bone_direction(which, static_cast<BoneDirection>((int)p_value));
+			} else if (opt == "length") {
+				set_end_bone_length(which, p_value);
+			} else {
+				return false;
+			}
+		} else if (what == "extend_end_bone") {
+			set_extend_end_bone(which, p_value);
+		} else if (what == "joint_count") {
+			set_joint_count(which, p_value);
+		} else if (what == "joints") {
+			int idx = path.get_slicec('/', 3).to_int();
+			String prop = path.get_slicec('/', 4);
+			if (prop == "bone_name") {
+				set_joint_bone_name(which, idx, p_value);
+			} else if (prop == "bone") {
+				set_joint_bone(which, idx, p_value);
+			} else {
+				return false;
+			}
+		} else {
+			return false;
+		}
+	}
+	return true;
+}
+
+bool ChainIK3D::_get(const StringName &p_path, Variant &r_ret) const {
+	String path = p_path;
+
+	if (path.begins_with("settings/")) {
+		int which = path.get_slicec('/', 1).to_int();
+		String what = path.get_slicec('/', 2);
+		ERR_FAIL_INDEX_V(which, (int)settings.size(), false);
+
+		if (what == "root_bone_name") {
+			r_ret = get_root_bone_name(which);
+		} else if (what == "root_bone") {
+			r_ret = get_root_bone(which);
+		} else if (what == "end_bone_name") {
+			r_ret = get_end_bone_name(which);
+		} else if (what == "end_bone") {
+			String opt = path.get_slicec('/', 3);
+			if (opt.is_empty()) {
+				r_ret = get_end_bone(which);
+			} else if (opt == "direction") {
+				r_ret = (int)get_end_bone_direction(which);
+			} else if (opt == "length") {
+				r_ret = get_end_bone_length(which);
+			} else {
+				return false;
+			}
+		} else if (what == "extend_end_bone") {
+			r_ret = is_end_bone_extended(which);
+		} else if (what == "joint_count") {
+			r_ret = get_joint_count(which);
+		} else if (what == "joints") {
+			int idx = path.get_slicec('/', 3).to_int();
+			String prop = path.get_slicec('/', 4);
+			if (prop == "bone_name") {
+				r_ret = get_joint_bone_name(which, idx);
+			} else if (prop == "bone") {
+				r_ret = get_joint_bone(which, idx);
+			} else {
+				return false;
+			}
+		} else {
+			return false;
+		}
+	}
+	return true;
+}
+
+void ChainIK3D::get_property_list(List<PropertyInfo> *p_list) const {
+	String enum_hint;
+	Skeleton3D *skeleton = get_skeleton();
+	if (skeleton) {
+		enum_hint = skeleton->get_concatenated_bone_names();
+	}
+
+	LocalVector<PropertyInfo> props;
+
+	for (uint32_t i = 0; i < settings.size(); i++) {
+		String path = "settings/" + itos(i) + "/";
+		props.push_back(PropertyInfo(Variant::STRING, path + "root_bone_name", PROPERTY_HINT_ENUM_SUGGESTION, enum_hint));
+		props.push_back(PropertyInfo(Variant::INT, path + "root_bone", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_NO_EDITOR));
+		props.push_back(PropertyInfo(Variant::STRING, path + "end_bone_name", PROPERTY_HINT_ENUM_SUGGESTION, enum_hint));
+		props.push_back(PropertyInfo(Variant::INT, path + "end_bone", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_NO_EDITOR));
```

Comment:
Does this serialize both bone_name and bone (index)? I think we should pick one. We should only serialize bone_name.

Reply:
Using both the bone name and index inherits the design of BoneAttachment. All other SkeletonModifier bone settings are serialized in the same way.

I remember discussing this with @fire in the past, and the main reasons for this design were to maintain compatibility with BoneAttachment and prevent glTF reimporting from breaking the indices (by making priority with the names).

It might be possible to remove this in the future, like in Godot5, but we should maintain consistency at least throughout Godot4. Therefore, this should not be changed in this PR.

If we want to change this, we should send a PR later that changes all other Nodes for Skeleton at the same time.